### PR TITLE
Parallelization

### PR DESCRIPTION
**Pull request by Jan-Williams**
Still need to update sparse matrix initialization to not hog memory, but I think that probably makes sense to do on the merge with the eigval solver.

### CODE REVIEW

File: src/orc/models/esn.py [hunk 1]
```diff
@@ -53,6 +56,9 @@ def __init__(
         Wr_spectral_radius: float = 0.8,
         dtype: type = jnp.float64,
         seed: int = 0,
+        chunks: int = 1,
+        locality: int = 0,
+        quadratic: bool = False
```
Comment (dtretiak) [inline src/orc/models/esn.py]: I think this is fine for now, but down the line would rather have a input that is a list of functions that are then applied to the reservoir states. Quadratic would then look like `nonlin_transforms = [lambda r: r, lambda r: r**2]` or something similar. And then we can implement a `NonlinearReadout` class that deals with this

File: src/orc/models/esn.py [hunk 2]
```diff
@@ -90,21 +103,132 @@ def __init__(
...
+            res_dim=res_dim,
+            in_dim=data_dim,
+            out_dim=data_dim,
             dtype=dtype,
             seed=seed,
         )
+
+    @eqx.filter_jit
+    def forecast(self, fcast_len: int, res_state: Array) -> Array:
+        """Forecast from an initial reservoir state.
```
Comment (dtretiak) [inline src/orc/models/esn.py:144]: I'm confused why the forecasting function was taken out of the `ReservoirComputerBase` class. In principle, all RC models should have the a forecasting function so it makes sense to put it in base. I envision it will also be unchanged from this form for most models, and if it really needs to be different, it can be overridden.

File: src/orc/models/esn.py (hunk 1 above)
Comment (Jan-Williams) [reply src/orc/models/esn.py]: Yeah, I'm remembering us talking about this now. I agree that's the better way forward in the long run. We need to think about how we want to handle which entries in the reservoir state are modified by the nonlinearity, though. Perhaps for another PR though?

File: src/orc/models/esn.py (hunk 2 above)
Comment (Jan-Williams) [reply src/orc/models/esn.py:144]: My thoughts on this depend a lot on what we decide to do with adding classifiers and other RC based models that need more information (i.e. control inputs) to "forecast." I'm good with pushing forecast back into `ReservoirComputerBase` but I think if we do we should change the name to `RCForecaster` or something like that.

File: src/orc/embeddings.py [hunk 3]
```diff
@@ -135,17 +153,69 @@ def __init__(
...
+            raise ValueError(
+                "Only 1-dimensional localization is currently supported, detected a "
+                f"{len(in_state.shape)}D field."
+            )
+        aug_state = jnp.hstack(
+            [in_state[-self.locality :], in_state, in_state[: self.locality]]
+        )
+        if not periodic:
+            aug_state = aug_state.at[: self.locality].set(aug_state[self.locality])
+            aug_state = aug_state.at[-self.locality :].set(aug_state[-self.locality])
```
Comment (dtretiak) [inline src/orc/embeddings.py]: Do we have tests for both periodic and non-periodic? I'm somewhat confused by how the non-peridic case is handled. I'm assuming this just removes overlaps for the boundary inputs?

File: src/orc/drivers.py [hunk 4]
```diff
@@ -145,55 +158,97 @@ def __init__(
...
         self.dtype = dtype
         key = jax.random.key(seed)
-        if spec_rad <= 0:
+        if spectral_radius <= 0:
             raise ValueError("Spectral radius must be positve.")
         if leak < 0 or leak > 1:
             raise ValueError("Leak rate must satisfy 0 < leak < 1.")
         if density < 0 or density > 1:
             raise ValueError("Density must satisfy 0 < density < 1.")
         wrkey1, wrkey2 = jax.random.split(key, 2)
```
Comment (dtretiak) [inline src/orc/drivers.py]: unused?

File: src/orc/drivers.py [hunk 5]
```diff
@@ -145,55 +158,97 @@ def __init__(
...
+                self.wr, res_state, proj_vars, self.bias * jnp.ones_like(proj_vars)
+            )
+            + (1 - self.leak) * res_state
         )
-        res_next = self.leak * res_next + (1 - self.leak) * res_state
-        return res_next
+
+    @staticmethod
+    @sparse.sparsify
+    @jax.vmap
```
Comment (dtretiak) [inline src/orc/drivers.py:220]: Have you seen concrete performance improvements from using @sparse.sparsify? Since `wr` is already sparse, doesn't `@` just invoke the sparse mat-vec mulitiplication anyways? Do we get extra performance out of this? I'm mainly being cautious since the transform seems quite experimental from the jax side and I'm worried about it changing on us

File: src/orc/models/esn.py (hunk 1 above)
Comment (dtretiak) [reply src/orc/models/esn.py]: yeah let's leave it for another PR. I think this is a good first step in that direction

File: src/orc/models/esn.py (hunk 2 above)
Comment (dtretiak) [reply src/orc/models/esn.py:144]: Ok I would agree with making an `RCForecasterBase` and later an `RCClassifierBase` . I think having them separate may also help with delineating training functions down the line as well

Comment (dtretiak) [reply src/orc/models/esn.py:144]: In order to avoid muddying up the commit history, I'm thinking we commit it as is. But then soon after, rename the base class and place the forecasting function back in the base with a new PR. sound good? Later down the line, we can make a classifier branch

Comment (Jan-Williams) [reply src/orc/models/esn.py:144]: That seems reasonable to me.

File: src/orc/drivers.py (hunk 5 above)
Comment (Jan-Williams) [reply src/orc/drivers.py:220]: The issue isn't performance, it's that vmap-ing over batch dimensions in Jax sparse matrices requires the use of the sparsify transform. So I think we are stuck with `@sparse.sparsify`.

File: src/orc/embeddings.py (hunk 3 above)
Comment (Jan-Williams) [reply src/orc/embeddings.py]: Non-periodic case just pads either side of the array with constant values equal to the start/end of the original array. To keep dimensions consistent for parallelization, we need to do some sort of padding and this seemed to be a reasonable approach. Zero padding would also be an option.

File: src/orc/drivers.py (hunk 4 above)
Comment (Jan-Williams) [reply src/orc/drivers.py]: Fixed in subsequent commit.

File: src/orc/embeddings.py (hunk 3 above)
Comment (dtretiak) [reply src/orc/embeddings.py]: I'm curious if its strictly necessary to have consistent dimensions for parallelization. I.e. each interior chunk would roughly have size `int(in_dim/chunks) + 2*locality` and boundary chunks would just have size `int(in_dim/chunks) + locality`. Regardless I think this is somewhat minor for now, and can be addressed in a later PR

File: src/orc/drivers.py (hunk 5 above)
Comment (dtretiak) [reply src/orc/drivers.py:220]: ah makes sense
